net: lwm2m: Implement gt/lt/st attributes handling

rlubos opened the pull request:
This PR implements Greater Than / Less Than / Step attributes handling in the LwM2M client library. While it was already possible to specicfy those attributes for resources, they were ignored by the LwM2M observe engine.

The biggest challenge was to decide on how to store the last notified value for resources that have gt/lt/st attributes configured. I propose a simple regsitry where an entry is allocated whenever one of the gt/lt/st attributes is set on the resource, similarily to how the actual attributes are stored. This approach seemed the best apporach memory-wise, although it obviously additionally limits the maximum number of resources that can have one of those numerical attributes set.

I'm not adding any tests for the feature, as the interop test spec doesn't actually seem to specify any tests to verify gt/lt/st attributes behavior and testing this in ztest didn't really seem viable.

Resolves #79306

rlubos committed:
net: lwm2m: Do not apply gt/lt/st attributes to incompatible resources

gt/lt/st attributes can only be applied to numerical resources according
to LwM2M specification (and common sense), hence verify that when
attributes are assigned.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Implement notified value registry for gt/lt/st attributes

Greater than, Less than and Step attributes require to track the last
notified value of a resource/resource instance in order to be able to
apply the specific thershold/step rules that the attributes define.

Therefore, implement a simple registry of the last notified values. When
one of the gt/lt/st attributes is configured on a resource/resource
instance, a registry entry is allocated for respective resource.
Whenever a notification is sent (either as a reply to Observe message,
or proactively) the registry is updated for the notified resources.

The stored resource values have been unified as "double" variables, to
avoid implementation complexity of having to support different
integer/floating point resource types.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Update value of resources in notification registry

Update the value stored in the notification registry whenever
notification is sent (for those resources that have gt/lt/st attributes
assigned).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Implement gt/lt/st attributes handling

While it was possible to specify Greater Than / Less Than / Step
observe attributes for a resource, they were not taken into
consideration when evaluating notification criteria. This commit adds
support for checking if the resource value meets the value criteria
specified by gt/lt/st attributes.

Note that gt/lt attributes are thresholds - notification should only
sent if the resource value crosses them.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Prevent busy looping when active notify is set

In case a new notification is triggered while the previous one is still
active (i.e. not acknowledged), the LwM2M engine will start busy
looping, as the new notification is still valid (and supposed to be sent
after the previous one is done), but it's not sent in the current
iteration. The busy looping might prevent the system from receiving the
acknowledgment for the previous one, resulting in unresponsive system.

Fix this, by delaying the new notification timestamp a bit in such case,
so that the new notification will be scheduled for later instead of
trying to send it again right away.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

kartben commented on subsys/net/lib/lwm2m/lwm2m_observation.c:
```diff
@@ -1643,3 +1990,146 @@ void lwm2m_engine_clear_duplicate_path(sys_slist_t *lwm2m_path_list, sys_slist_t
+	}
+
+	if (!res->multi_res_inst) {
+		if (res->res_instances > 0) {
```
```suggestion
		if (res->res_inst_count > 0) {
```

kartben commented on subsys/net/lib/lwm2m/lwm2m_observation.c:
```diff
@@ -87,12 +87,142 @@ static const uint8_t LWM2M_ATTR_LEN[] = {4, 4, 2, 2, 2};
+		notify_reg = notify_value_reg_alloc(ref);
+	}
+
+	if (notify_reg == NULL) {
```
Drop?

rlubos replied:
I've ended up keeping the if as I've also added a log, fixed the return statement below.

SeppoTakalo reviewed (changes requested):
I like this. Good work.

I left few remarks about too much logging and I believe there was a bug that Copilot pointed out.

Also, I would like to see some tests provided with this one as it updates core functionality of the engine.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_observation.c:
```diff
@@ -87,12 +87,142 @@ static const uint8_t LWM2M_ATTR_LEN[] = {4, 4, 2, 2, 2};
+
+	return 0;
+
+cleanup:
+	for (int i = 0; i < res->res_inst_count; i++) {
+		bool skip = false;
+
+		for (int j = 0; j < ARRAY_SIZE(write_attr_pool); j++) {
+			struct lwm2m_attr *attr = &write_attr_pool[i];
+
+			/* If register was allocated individually for the resource
+			 * instance earlier - skip.
+			 */
+			if (attr->ref == &res->res_instances[i] &&
+			    (attr->type == LWM2M_ATTR_LT ||
+			     attr->type == LWM2M_ATTR_GT ||
+			     attr->type == LWM2M_ATTR_STEP)) {
+				skip = true;
+				break;
+			}
+		}
+
+		(void)notify_value_reg_inst_update(&res->res_instances[i], false);
+	}
+
+	return -ENOMEM;
```
> Either we prevent registering attributes to resource level if resource instance level already has attributes.

Why would we do that? Spec doesn't forbid setting the same attribute at both res instance and resource level, moreover, it's states clearly that resource instance attribute takes preference in such case.

> Or we clean up all resource instance allocations when we run out or mem. Possibly releasing previously allocated parameters.

We could do that, but it has a built-in design flaw of clearing any previously allocated registry entries at the instance level. And to be honest, the clenaup logic is not that complex, just loop over attributes searching for resource instances.
If you insist, I can simplify it to remove entries for **all** instances, but I'm not sure if it's worth to introduce such limitation just to get rid of the inner loop.

>When we run out of mem, we clearly indicate that, (maybe LOG_ERR would also be OK here) which gives the developer change to fix the logic or allocate more entries.
> Also, fix the bug pointed out by Github Copilot (using index i instead of j).

Copilot bug fixed, log message added.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_observation.c:
```diff
@@ -1643,3 +1990,146 @@ void lwm2m_engine_clear_duplicate_path(sys_slist_t *lwm2m_path_list, sys_slist_t
+		return;
+	}
+
+	if (!res->multi_res_inst) {
+		if (res->res_instances > 0) {
+			res_inst = res->res_instances;
+
+			return update_resource_instance_notified_value(path, res_inst);
+		}
+	}
```
I think you're right, I thought optional resources would be a problem but we always seem to set the res_instances pointer anyway... I'll drop this part and test a bit to see if it works well.

rlubos commented:
> Also, I would like to see some tests provided with this one as it updates core functionality of the engine.

@SeppoTakalo How to test such a fucntionality w/o a LwM2M server though, any ideas? The interop tests does not specify any test cases for these attributes, and I don't really see a ztest tests for this as we don't have any server to communicate with.

SeppoTakalo commented:
> @SeppoTakalo How to test such a fucntionality w/o a LwM2M server though, any ideas? The interop tests does not specify any test cases for these attributes, and I don't really see a ztest tests for this as we don't have any server to communicate with.

True, there seems to be no ztests for attribute handling.

On Pytest interoperability tests, you could implement tests for these.
Writing attributes are already implemented, see test case 261.
Observation is tested in testcase 301.

So if you copy a testcase 301 remove `pmin` and add `step` or any of the attributes you implemented, you could test from the timestamp if the Notify was coming immediately (because the step was triggered) or if was coming because of `pmax`.

As this testcase is not in the standard, you can create a new file for it, like `test_obs_attributes.py`

rlubos committed:
tests: net: lwm2m: interop: Add custom tests for observe atributes

Add custom interoperability tests verifying that numerical observe
attributes work as expected.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
